crates/blind_pool/src/pool.rs

```diff
@@ -0,0 +1,563 @@
+/// the items it contains, except when explicitly called to operate on an item (e.g. `remove()`
+/// implies exclusive access).
+///
+/// # Resource usage
+///
+/// The collection automatically grows as items are added. To reduce memory usage after items have
+/// been removed, use the [`shrink_to_fit()`][1] method to release unused capacity.
+///
+/// [1]: Self::shrink_to_fit
+///
+/// # Example
+///
+/// ```rust
+/// use blind_pool::BlindPool;
+///
+/// let mut pool = BlindPool::new();
+///
+/// // Insert values of different types.
+/// let pooled_u32 = pool.insert(42_u32);
+/// let pooled_i64 = pool.insert(-123_i64);
+///
+/// // Read from the memory.
+/// // SAFETY: The pointers are valid and the memory contains the values we just inserted.
+/// let value_u32 = unsafe { pooled_u32.ptr().read() };
+/// let value_i64 = unsafe { pooled_i64.ptr().read() };
+///
+/// assert_eq!(value_u32, 42);
+/// assert_eq!(value_i64, -123);
+/// ```
+#[derive(Debug)]
+pub struct BlindPool {
+    /// Internal pools, one for each unique memory layout encountered.
+    /// We use foldhash for better performance with small hash tables.
+    pools: HashMap<Layout, OpaquePool, FixedState>,
+
+    /// Drop policy that determines how the pool handles remaining items when dropped.
+    drop_policy: DropPolicy,
+}
+
+impl BlindPool {
+    /// Creates a new [`BlindPool`] with default configuration.
+    ///
+    /// This is equivalent to [`BlindPool::builder().build()`][BlindPool::builder].
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use blind_pool::BlindPool;
+    ///
+    /// let mut pool = BlindPool::new();
+    ///
+    /// let pooled = pool.insert(42_u64);
+    ///
+    /// // SAFETY: The pointer is valid and contains the value we just inserted.
+    /// let value = unsafe { pooled.ptr().read() };
+    /// assert_eq!(value, 42);
+    /// ```
+    #[must_use]
+    pub fn new() -> Self {
+        Self::builder().build()
+    }
+
+    /// Creates a builder for configuring and constructing a [`BlindPool`].
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use blind_pool::{BlindPool, DropPolicy};
+    ///
+    /// let pool = BlindPool::builder()
+    ///     .drop_policy(DropPolicy::MustNotDropItems)
+    ///     .build();
+    /// ```
+    pub fn builder() -> BlindPoolBuilder {
+        BlindPoolBuilder::new()
+    }
+
+    /// Creates a new [`BlindPool`] with the specified configuration.
+    ///
+    /// This method is used internally by the builder to construct the actual pool.
+    #[must_use]
+    pub(crate) fn new_inner(drop_policy: DropPolicy) -> Self {
+        Self {
+            pools: HashMap::with_hasher(FixedState::default()),
+            drop_policy,
+        }
+    }
+
+    /// Inserts a value into the pool and returns a handle to access it.
+    ///
+    /// The pool automatically determines the memory layout of the type and uses or creates
+    /// an appropriate internal pool to store the value.
```

Comment:
The API documentation you created still has numerous references to implementation details like "internal pools". Such references are inappropriate - stick to only documenting the public API contract in the API documentation of public types.